haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,415 @@
+    the pipeline as is.
+
+    Shaper is especially useful in the context of pipelines with PromptNode(s) where we need to modify the invocation
+    context to match the templates of PromptNodes.
+
+    Multiple Shaper components can be used in a pipeline to modify the invocation context as needed.
+
+    The following functions are supported by Shaper:
+
+    - len: parameters are target:Any; returns int; built-in len function returning the length of the input variable
+    - expand: parameters are [expand_target:Any, size:int]; returns List[Any]; it returns a list of specified size
+        for the input variable
+    - concat: parameters are [texts:List[str], delimiter:str]; returns str, concatenates texts with the specified
+        delimiter
+    - concat_docs: parameters are [docs: List[Document]], delimiter:str]; returns str; concatenates the docs with the
+        specified delimiter and returns a string
+    - convert_to_docs: parameters are texts: List[str]; returns List[Document]; converts the input list of strings
+        to a list of Documents
+    """
+
+    outgoing_edges = 1
+
+    def __init__(self, inputs: Dict[str, Any]):
+        """
+        Initializes a Shaper component.
+
+        :param inputs: A dictionary of input parameters for the Shaper component. These directives
+        are a dictionary version of YAML directives that specify the functions to invoke on the invocation context.
+
+        For example, in the Python snippet below:
+        ```python
+            shaper = Shaper(inputs={"query": {"output": "questions"}})
+            pipeline = Pipeline()
+            pipeline.add_node(component=shaper, name="shaper", inputs=["Query"])
+            ...
+        ```
+        Shaper component is initialized with a directive to rename the invocation context variable query to questions.
+        """
+        super().__init__()
+        self.inputs = inputs
+
+        concat: Callable[[List[str], str], str] = lambda docs, delimiter=" ": delimiter.join(docs)  # type: ignore
+        concat_docs: Callable[[List[Document], str], str] = lambda docs, delimiter=" ": delimiter.join(  # type: ignore
+            [d.content for d in docs]
+        )
+        convert_to_docs: Callable[[List[str]], List[Document]] = lambda texts: [Document(text) for text in texts]
+        tmp_registry: Dict[str, Callable] = {
+            "len": len,
+            "expand": (lambda expand_target, size: [expand_target] * size),
+            "expand:size": lambda documents: len(documents),  # pylint: disable=unnecessary-lambda
+            "concat": concat,
+            "concat_docs": concat_docs,
+            "convert_to_docs": convert_to_docs,
+        }
+        # convert tmp_registry to a read-only immutable dictionary for added security
+        # see https://adamj.eu/tech/2022/01/05/how-to-make-immutable-dict-in-python/
+        self.function_registry = MappingProxyType(tmp_registry)
```

Comment:
True, what if we use some of the ideas from this SO [post](https://stackoverflow.com/questions/32760802/how-to-write-metaclass-which-would-prevent-creating-new-attributes-after-init). It seems possible.

Reply:
In Python we can't get away with this, if anything we're worsening our security posture by making the code more complex - let's just use a dict.

Reply:
We're just moving the problem up :smile: Don't forget that if someone has access to the code to the point that they can set attributes it's already game over, because they can just make their own custom nodes, subclass this node, remove the metaclasses with monkeypatching... you can't win this game in Python. Trying to make classes immutable is really hard and there's always a way around it, by design (by the way, in that same SO thread the commenters note exactly this).

What is important is that a person can't inject functions through the YAML definition. Everything below that is not going to help much imho.